include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -0,0 +1,359 @@
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_ALGORITHM_RANGES_IMPL_H
+#define _ONEDPL_ALGORITHM_RANGES_IMPL_H
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+
+#include <ranges>
+#include "algorithm_fwd.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+namespace __ranges
+{
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_for_each
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+void
+__pattern_for_each_impl(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto __f_1 = 
+        [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+
+    oneapi::dpl::__internal::__pattern_walk1(__tag, std::forward<_ExecutionPolicy>(__exec), std::ranges::begin(__r),
+        std::ranges::begin(__r) + std::ranges::size(__r), __f_1);
+}
+
+template <typename _IsVector, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+void
+__pattern_for_each(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+}
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+void
+__pattern_for_each(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+    else
+        std::ranges::for_each(std::forward<_R>(__r), __f, __proj);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_transform
+//---------------------------------------------------------------------------------------------------------------------
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F, typename _Proj>
+void
+__pattern_transform_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _F __op,
+                         _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+    assert(std::ranges::size(__in_r) == std::ranges::size(__out_r));
+
+    auto __unary_op = [=](auto&& __val) -> decltype(auto) { 
+        return std::invoke(__op, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
```

Comment:
Maybe this is nit-picking, but I think we are duplicating the implementation of this, and maybe a few other of these similar lambdas.  The previous instance I'm looking at is in `__pattern_for_each_impl`.  

Would we be better off with these as defined `struct` callable helpers to standardize and share our implementations between patterns?  

One side effect is that it would be reasonably easy to add device-copyable specializations for these (lambdas currently can't be specialized in the same way).  We are creating a lot more exposure for device copyable problems via lambas with all the projection operators in this implementation (if any projection uses a std::tuple for instance).

This could be possibly improved in a follow-up PR.

Reply:
1) Yes, we can consider follow-up refactoring to minimize code duplication.
2) "Device copyable" concerns are not actual here because the file contains only host-executed code.

Reply:
(1) Ok, I can make an issue, and resolve these comments for now.  
(2) Yes, you are correct, but the same comment applies for the hetero implementation, I just missed up the placement.

Reply:
resolving due to created issue #1799